File: dashboards/ruby/routes/api_keys.rb
# frozen_string_literal: true

# API Keys routes — list, create, revoke

get '/api-keys' do
  require_auth!
  res = api_get('/api/api-keys')
  @keys = res[:status] == 200 ? (res[:body]['api_keys'] || res[:body]['keys'] || res[:body]) : []
  @keys = @keys.is_a?(Array) ? @keys : []
  @created_key = session.delete(:created_key)
  erb :api_keys
end

post '/api-keys' do
  require_auth!
  res = api_post('/api/api-keys', {
    name: params[:name],
    scopes: params[:scopes].to_s.split(',').map(&:strip).reject(&:empty?)
  })
  if res[:status] < 300
    key = res[:body]['key'] || res[:body]['api_key'] || res[:body]['token']
    if key
      session[:created_key] = key
      set_flash('API key created. Copy it now -- it will not be shown again.', 'success')
    else
      set_flash('API key created', 'success')
    end
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to create API key') : 'Failed to create API key', 'danger')
  end
  redirect '/api-keys'
end

post '/api-keys/:id/revoke' do
  require_auth!
  res = api_delete("/api/api-keys/#{params[:id]}")
  if res[:status] < 300
    set_flash('API key revoked successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to revoke key') : 'Failed to revoke key', 'danger')
  end
  redirect '/api-keys'
end

File: dashboards/python/routes/api_keys.py
"""
API Keys routes: list, create, and revoke API keys.
"""

from flask import Blueprint, render_template, request, session, redirect, url_for, flash
from functools import wraps
from utils.api import api_request

api_keys_bp = Blueprint('api_keys_bp', __name__)


def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'token' not in session:
            return redirect(url_for('auth_bp.login'))
        return f(*args, **kwargs)
    return decorated


@api_keys_bp.route('/api-keys')
@login_required
def api_keys():
    data = api_request('GET', '/api/api-keys')
    return render_template(
        'api_keys.html',
        active_page='api_keys',
        user=session.get('user', {}),
        keys=data.get('keys', []),
    )


@api_keys_bp.route('/api-keys/create', methods=['POST'])
@login_required
def create_api_key():
    result = api_request('POST', '/api/api-keys', {'name': request.form['name']})
    if 'plaintext' in result:
        flash(f"Key created: {result['plaintext']} — SAVE THIS NOW!")
    else:
        flash(result.get('error', 'Failed'))
    return redirect(url_for('api_keys_bp.api_keys'))


@api_keys_bp.route('/api-keys/<id>/revoke', methods=['POST'])
@login_required
def revoke_api_key(id):
    api_request('DELETE', f'/api/api-keys/{id}')
    flash('Key revoked')
    return redirect(url_for('api_keys_bp.api_keys'))
